lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,246 @@
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, type)
+
+        mutex.synchronize do
+          if aggregation_state.key?(key) && aggregation_state[key][:value].size + 1 >= @max_values
+            do_flush
+          end
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = {
+              type: type,
+              name: name,
+              value: [],
+              tags: tags,
+              no_prefix: no_prefix,
+            }
+          end
+          aggregation_state[key][:value] << value
+        end
+      end
+
+      def gauge(name, value, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).g(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        mutex.synchronize do
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = {
+              type: GAUGE,
+              name: name,
+              value: value,
+              tags: tags,
+              no_prefix: no_prefix,
+            }
+          end
+          # keep the last value
+          aggregation_state[key][:value] = value
+        end
+      end
+
+      def flush
+        mutex.synchronize(&method(:do_flush))
+      end
+
+      private
+
+      def do_flush
+        aggregation_state.each do |_key, agg|
+          case agg[:type]
+          when COUNT
+            sink << datagram_builder(no_prefix: agg[:no_prefix]).c(
+              agg[:name],
+              agg[:value],
+              CONST_SAMPLE_RATE,
+              agg[:tags],
+            )
+          when DISTRIBUTION, MEASURE, HISTOGRAM
+            sink << datagram_builder(no_prefix: agg[:no_prefix]).distribution_value_packed(
+              agg[:name],
+              agg[:type].to_s,
+              agg[:value],
+              CONST_SAMPLE_RATE,
+              agg[:tags],
+            )
+          when GAUGE
+            sink << datagram_builder(no_prefix: agg[:no_prefix]).g(
+              agg[:name],
+              agg[:value],
+              CONST_SAMPLE_RATE,
+              agg[:tags],
+            )
+          else
+            StatsD.logger.error { "[#{self.class.name}] Unknown aggregation type: #{agg[:type]}" }
+          end
+        end
+        aggregation_state.clear
+      end
+
+      attr_reader :mutex, :aggregation_state, :sink
+
+      def tags_sorted(tags)
+        return [].freeze if tags.nil? || tags.empty?
+
+        if tags.is_a?(Hash)
+          sorted = tags.sort_by { |k, _v| k.to_s }
+          tags = sorted.map! { |k, v| "#{k}:#{v}" }
+        else
+          tags.sort!
+        end
+        tags
+      end
+
+      def packet_key(name, tags = [], no_prefix = false, type = COUNT)
+        "#{name}#{tags.join}#{no_prefix}#{type}".b
+      end
```

Comment:
If you're going to create a cache key, might as well do it in a way that allow you to re-use its element to generate the final packet.

e.g. tags should be formatted so that you can just interpolate them in the final packet, etc.

Reply:
Worth remembering @pedro-stanaka that anything in ruby can be a hash key, so for example:

```ruby
irb(main):001:0> cache = {}
=> {}
irb(main):002:0> tags = ["foo:bar", "a:b"].sort.join(",")
=> "a:b,foo:bar"
irb(main):003:0> key = ["some_name", tags, true, :c]
=> ["some_name", "a:b,foo:bar", true, :c]
irb(main):004:0> cache[key] = 1
=> 1
irb(main):005:0> cache[key] += 1
=> 2
irb(main):008:0> cache.keys.first
=> ["some_name", "a:b,foo:bar", true, :c]
```

(Technically, the object must be "hashable", but most things in ruby are. For something like an array or hash, the `.hash` method out of the box determines the hash in such a way as to be useful in this circumstance)

Reply:
Going to leave this for last. Would be nice do this, but I need to normalize and make sure my separator is not present in the final string parts. I also was thinking of sharing the method from `DatagramBuilder#compile_tags` or making it a class method so I can use it for this.